near/src/runtime.rs

```diff
@@ -864,66 +860,72 @@ impl RuntimeAdapter for NightshadeRuntime {
+                    part_id, num_parts, state_root, e
+                )
+                    .into()
+            })
+            .map(|a| a.try_to_vec().expect("Cannot serialize"))
     }
 
-    fn accept_state_part(
+    fn validate_state_part(
         &self,
         state_root: &StateRoot,
-        part: &StatePart,
-        _proof: &Vec<u8>,
-    ) -> Result<(), Box<dyn std::error::Error>> {
-        if part.part_id > 0 {
-            /* TODO MOO */
-            return Ok(());
+        part_id: u64,
+        num_parts: u64,
+        data: &Vec<u8>,
+    ) -> Result<bool, Error> {
+        if part_id >= num_parts {
+            return Err("Invalid part_id in validate_state_part".to_string().into());
         }
-        debug!(target: "runtime", "Writing state part #{} for shard #{} @ {}, size = {}", part.part_id, part.shard_id, state_root.hash, part.data.len());
-        // TODO prove that the part is valid
-        let mut state_update = TrieUpdate::new(self.trie.clone(), CryptoHash::default());
-        let state_part_len = part.data.len();
-        let mut cursor = Cursor::new(part.data.clone());
-        while cursor.position() < state_part_len as u64 {
-            let key_len = cursor.read_u32::<LittleEndian>()? as usize;
-            let mut key = vec![0; key_len];
-            cursor.read_exact(&mut key)?;
-            let value_len = cursor.read_u32::<LittleEndian>()? as usize;
-            let mut value = vec![0; value_len];
-            cursor.read_exact(&mut value)?;
-            state_update.set(key, DBValue::from_slice(&value));
+        match BorshDeserialize::try_from_slice(data) {
+            Ok(trie_nodes) => match Trie::validate_trie_nodes_for_part(
+                state_root,
+                part_id,
+                num_parts,
+                &trie_nodes,
+            ) {
+                Ok(_) => Ok(true),
+                // Storage error should not happen
+                Err(_) => Ok(false),
+            },
+            // Deserialization error means we've got the data from malicious peer
+            Err(_) => Ok(false),
         }
-        let (store_update, root) = state_update.finalize()?.into(self.trie.clone())?;
-        if root != state_root.hash {
-            return Err("Invalid state root".into());
+    }
+
+    fn confirm_state(&self, state_root: &StateRoot, data: &Vec<Vec<u8>>) -> Result<(), Error> {
+        let mut parts = vec![];
+        for part in data {
+            parts.push(
+                BorshDeserialize::try_from_slice(part)
+                    .expect("Part was already validated earlier, so could never fail here"),
+            );
         }
-        store_update.commit()?;
-        Ok(())
+        let trie_changes =
+            Trie::combine_state_parts(&state_root, &parts).expect("Failed to get trie update");
+        // TODO clean old states
+        let trie = self.trie.clone();
+        let (store_update, _) = trie_changes.into(trie).expect("Failed to apply trie update");
```

Comment:
Sounds like `TrieChanges::into` actually can never fail and needs refactoring, `.expect("TrieChanges::into never fails")`

Reply:
++